fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -458,4 +483,194 @@ public static long getFrom(@Nonnull LucenePartitionInfoProto.LucenePartitionInfo
     public static long getTo(@Nonnull LucenePartitionInfoProto.LucenePartitionInfo partitionInfo) {
         return Tuple.fromBytes(partitionInfo.getTo().toByteArray()).getLong(0);
     }
+
+    /**
+     * Re-balance full partitions, if applicable.
+     *
+     * @return void future
+     */
+    @Nonnull
+    public CompletableFuture<Void> rebalancePartitions() {
+        if (!isPartitioningEnabled()) {
+            return AsyncUtil.DONE;
+        }
+
+        // This function will iterate the grouping keys
+        final ScanProperties scanProperties = ScanProperties.FORWARD_SCAN.with(
+                props -> props.clearState().setReturnedRowLimit(1));
+
+        final Range range = state.indexSubspace.range();
+        final KeyRange keyRange = new KeyRange(range.begin, range.end);
+        final Subspace subspace = state.indexSubspace;
+        final KeyExpression rootExpression = state.index.getRootExpression();
+
+        if (! (rootExpression instanceof GroupingKeyExpression)) {
+            return processPartitionRebalancing(Tuple.from());
+        }
+
+        GroupingKeyExpression expression = (GroupingKeyExpression) rootExpression;
+        final int groupingCount = expression.getGroupingCount();
+
+        try (RecordCursor<Tuple> cursor = new ChainedCursor<>(
+                state.context,
+                lastKey -> FDBDirectoryManager.nextTuple(state.context, subspace, keyRange, lastKey, scanProperties, groupingCount),
+                Tuple::pack,
+                Tuple::fromBytes,
+                null,
+                ScanProperties.FORWARD_SCAN)) {
+
+            return cursor.map(tuple -> Tuple.fromItems(tuple.getItems().subList(0, groupingCount)))
+                    .forEachAsync(this::processPartitionRebalancing, 1);
+        }
+    }
+
+    /**
+     * Re-balance the first partition in a given grouping key by moving documents out of it.
+     *
+     * Note that in order to finish the task within the bounds of a single transaction, only the first
+     * partition needing re-balancing will be processed. If there are other partitions that need re-balancing,
+     * they will be processed during subsequent calls.
+     *
+     * @param groupingKey grouping key
+     * @return void future
+     */
+    @Nonnull
+    private CompletableFuture<Void> processPartitionRebalancing(@Nonnull final Tuple groupingKey) {
+
+        return getAllPartitionMetaInfo(groupingKey).thenCompose(partitionInfos -> {
```

Comment:
Instead of having to get the metadata entries of all partitions (in order to determine the max partition id, and the document count of each partition), perhaps an optimization would be to create a "summary" entry that has the max id as well as a map of partition metadata key -> doc count.
Thoughts?